core/src/main/java/org/apache/calcite/rel/rules/materialize/MaterializedViewRule.java

```diff
@@ -1252,6 +1285,193 @@ protected RexNode replaceWithOriginalReferences(final RexBuilder rexBuilder,
+        }
+      }
+      return modifiedVal;
+    }
+
+    /**
+     * Generates a predicate that is added to the view so that we can use union rewriting.
+     *
+     * @param call the query predicate
+     * @return the predicate that is to be used as the view predicate or a view filter
+     */
+    @SuppressWarnings("BetaApi")
+    @Override public RexNode visitCall(RexCall call) {
+      switch (call.getKind()) {
+      case SEARCH:
+        Sarg sarg = castNonNull(((RexLiteral) call.getOperands().get(1)).getValueAs(Sarg.class));
+        if (sarg.rangeSet.asRanges().size() != 1) {
+          return super.visitCall(call);
+        }
+        Range r = (Range) sarg.rangeSet.asRanges().iterator().next();
+        final int floorIndex = ((RexInputRef) floorCall.getOperands().get(0)).getIndex();
+        RexNode tableInputRefOperand = call.operands.get(0);
+        if (!(tableInputRefOperand instanceof RexTableInputRef)) {
+          // ignore SARG that isn't on table column
+          return rexBuilder.makeLiteral(true);
+        }
+        int predicateIndex = ((RexTableInputRef) tableInputRefOperand).getIndex();
+        if (floorIndex != predicateIndex) {
+          // ignore predicates on different columns
+          return rexBuilder.makeLiteral(true);
+        }
+        if (!(r.lowerEndpoint() instanceof TimestampString)
+            || !(r.upperEndpoint() instanceof TimestampString)) {
+          // ignore predicates where the type isn't a timestamp
+          return rexBuilder.makeLiteral(true);
+        }
+        RexNode lb = rexBuilder.makeTimestampLiteral((TimestampString) r.lowerEndpoint(), 0);
+        // if the lower bound is exclusive we need to shift the modified value to the next higher
+        // floored value
+        long modifiedLbVal = getModifiedVal(r.lowerBoundType() == BoundType.OPEN, true, lb);
+        TimestampString lbTimestampString = TimestampString.fromMillisSinceEpoch(modifiedLbVal);
+        RexNode ub = rexBuilder.makeTimestampLiteral((TimestampString) r.upperEndpoint(), 0);
+        // if the upper bound is inclusive we need to shift the modified value to the previous lower
+        // floored value
+        long modifiedUbVal = getModifiedVal(r.upperBoundType() == BoundType.CLOSED, false, ub);
+        TimestampString ubTimestampString = TimestampString.fromMillisSinceEpoch(modifiedUbVal);
+
+        lowerBound = modifiedLbVal;
+        upperBound = modifiedUbVal;
+
+        if (generateViewFilter) {
+          tableInputRefOperand = rexInputRef;
+        }
+
+        RelDataType type = call.getOperands().get(1).getType();
+        Sarg modSarg = Sarg.of(RexUnknownAs.UNKNOWN,
+            ImmutableRangeSet.of(Range.closedOpen(lbTimestampString, ubTimestampString)));
+        RexNode searchArgumentLiteral = rexBuilder.makeSearchArgumentLiteral(modSarg, type);
+        RexNode res = rexBuilder.makeCall(SqlStdOperatorTable.SEARCH, tableInputRefOperand,
```

Comment:
What about dropping `res` and returning the expressions?

Reply:
done